### PR TITLE
[gatsby-source-filesystem] Add Directory nodes + "relativeDirectory" field

### PR DESCRIPTION
**Pull request by thisiscam**
#2053

### CODE REVIEW

File: packages/gatsby-source-filesystem/src/gatsby-node.js [hunk 1]
```diff
@@ -35,28 +35,35 @@ exports.sourceNodes = (
     return Promise.all(queue.map(createAndProcessNode))
   }
 
-  watcher.on(`add`, path => {
```
Comment (KyleAMathews) [inline packages/gatsby-source-filesystem/src/gatsby-node.js]: Can you remove this refactor? I find named functions like this less clear.

File: packages/gatsby-source-filesystem/src/create-file-path.js [hunk 2]
```diff
@@ -7,7 +7,6 @@ function findFileNode({ node, getNode }) {
 
   let whileCount = 0
   while (
-    fileNode.internal.type !== `File` &&
```
Comment (KyleAMathews) [inline packages/gatsby-source-filesystem/src/create-file-path.js]: Why was this removed?

File: packages/gatsby-source-filesystem/src/create-file-node.js [hunk 3]
```diff
@@ -20,8 +21,24 @@ exports.createFileNode = async (pathToFile, pluginOptions = {}) => {
     absolutePath: slashed,
   }
   // console.log('createFileNode', slashedFile.absolutePath)
-  const contentDigest = await md5File(slashedFile.absolutePath)
   const stats = await fs.stat(slashedFile.absolutePath)
+  const isDirectory = stats.isDirectory()
+  var internal;
+  if(isDirectory) {
+    const contentDigest = await md5Dir(slashedFile.absolutePath)
```
Comment (KyleAMathews) [inline packages/gatsby-source-filesystem/src/create-file-node.js]: This isn't necessary. It's not interesting to know for the directory node if a file within the directory changed. Instead you should make a digest of the directory stats object and anything else we add to the Directory node.

File: packages/gatsby-source-filesystem/src/create-file-node.js [hunk 4]
```diff
@@ -15,13 +16,41 @@ exports.createId = createId
...
 exports.createFileNode = async (pathToFile, pluginOptions = {}) => {
   const slashed = slash(pathToFile)
+  const parsedSlashed = path.parse(slashed)
   const slashedFile = {
-    ...path.parse(slashed),
+    ...parsedSlashed,
     absolutePath: slashed,
+    // Useful for limiting graphql query with certain parent directory
+    dirRelative: path.relative(
+                  pluginOptions.path || process.cwd(),
```
Comment (KyleAMathews) [inline packages/gatsby-source-filesystem/src/create-file-node.js]: There's already "relativePath". Perhaps just name this "relativeDirectory"?

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit d99d2ce8cd9cdf778af066deeef5220835451cc0

https://app.netlify.com/sites/using-glamor/deploys/59dfc97da6188f2cc5cc9399

**Comment by KyleAMathews** [comment]
Thanks @thisiscam!

**Comment by jlengstorf** [comment]
Hiya @thisiscam! 👋

This is definitely late, but on behalf of the entire Gatsby community, I wanted to say thank you for being here.

Gatsby is built by awesome people like you. Let us say “thanks” in two ways:

1.  **We’d like to send you some Gatsby swag.** As a token of our appreciation, you can go to the [Gatsby Swag Store][store] and log in with your GitHub account to get a coupon code good for one free piece of swag. (We’ve got t-shirts and hats, plus some socks that are really razzing our berries right now.)
2.  **If you’re not already part of it, we just invited you to join the Gatsby organization on GitHub.** This will add you to our team of maintainers. You’ll receive an email shortly asking you to confirm. By joining the team, you’ll be able to label issues, review pull requests, and merge approved pull requests.

If you have questions, please don’t hesitate to reach out to us: tweet at [@gatsbyjs][twitter] and we’ll come a-runnin’.

Thanks again! 💪💜

[store]: https://store.gatsbyjs.org
[twitter]: https://twitter.com/gatsbyjs
